### ISSUE TITLE
PPO very different performance compared to StableBaselines3

### ISSUE THREAD

**Issue by cantor-dust** [AIRLab-POLIMI/mushroom#81]
Hi,

I have a question regarding the performance of PPO in MushroomRL compared to StableBaselines3.

Is it safe to say that your implementation of PPO should achieve a similar performance (in terms of mean discounted reward) compared to the PPO implementation of  StableBaselines3?

By looking at the code it seems sensible to compare the two performances, by setting some hyper-parameters of PPO from StableBaselines3 in order to be able to compare it with your implementation.

However when running some experiments I came across something different:
1. Running experiments on MushroomRL InvertedPendulum environment, your implementation achieves superior performance (in terms of mean discounted reward). 

2. Running experiments on an LQG environment, StableBaselines3 reaches the theoretical optimal policy whereas your implementation is extremely far off.

In both cases I used the same number of steps, the same hyper-parameters, the same network architecture, the same policy, the same environment and the same evaluation method.

Are you surprised by this? If not, why? 

Thank you in advance

**Comment by davide-tateo90** [comment]
If I'm not wrong, stable baselines3 implements the PPO2 version of the algorithm from openai (not stable) baselines. Our version matches PPO 1.

One clear difference is the fit of the value function, which is done separately from the policy update.
Other differences may be input normalization and different handling of the terminal states (we differentiate from truncated episodes and absorbing states, and we handle this correctly in the algorithm).

Are you sure that you used the same network? the stable baselines 3 may use a shared network implementation or different activation functions. In mushroom it's trivial to check the difference, while in stable baselines may not be so easy.

**Comment by cantor-dust** [comment]
Yes I use the same network: I can specify the number of layers, the number of neurons, the activation function and whether to use a shared network. This can also be checked by extracting the policy from the model object in StableBaselines3. I also use the same initial variance for the policy.

The biggest factor that messes up the result achieved by your implementation is that the final variance tensor has components which are still quite big, whereas in StableBaselines3 the final variance tensor has smaller components.
But I used the same initial variance in both cases.

Since your implementation matches that of PPO1, then something else may cause this issue. Could it have something to do with the policy? (i.e: with how you perform exploration?)

**Comment by davide-tateo90** [comment]
the policy depends on the policy you choose.
If you use the gaussian policy, the variance is parametrized by log std. See here:
[GaussianTorchPolicy](https://github.com/MushroomRL/mushroom-rl/blob/62ff8ab7f1ac886d415b53a805a60175f3397071/mushroom_rl/policy/torch_policy.py#L181)

Not sure if stable baselines use the same policy, but if you see our benchmarks, the performance matches the ones on the paper [see here](https://mushroom-rl-benchmark.readthedocs.io/en/latest/source/benchmarks/actor_critic/mujoco.html)

**Comment by cantor-dust** [comment]
Yeah in both tests I am using a Gaussian Policy. 

Even if there were slight modifications between the two implementations, still I would expect both implementations to be able to solve a very simple LQG environment.

**Comment by davide-tateo90** [comment]
Unfortunately, I cannot give support on stable baselines.
Regarding our implementation, as you see it matches the one from the paper. 
I'm quite sure it can solve the LQG environment if you change the tuning of the hyperparameters. Not knowing how they are implemented in stable baselines 3, I can only suppose that there may be different scales/usage/meaning or different input normalization (maybe reward normalization that is not present in mushroom), or maybe a different training (mixing the value function with the policy on the training step might have a scale effect on the optimization).

The implementation of mushroom is quite straightforward to check, and the results on the benchmark are obtained with the current implementation (except, maybe, updates on NumPy/torch libraries)

**Comment by cantor-dust** [comment]
I found the issue, it was a bug on my end...
Thank you for your help, and sorry I wasted your time.
